File: PythonSolutions/Neetcode150/pointer_problems/11_container-with-most-water.py
from typing import List
class Solution:
    def maxArea(self, height: List[int]) -> int:
        l, r = 0, len(height) - 1
        maxArea = 0

        while l < r:
            # Get min height so water doesn't overflow
            area = min(height[l], height[r]) * (r-l)
            maxArea = max(maxArea, area)

            # If the height on the left is smaller then find a new height for left
                # We want to maximize the highest chance of finding the highest combinations of height
            if height[l] < height[r]:
                l += 1
            else:
                r -= 1
        
        return maxArea

File: TypescriptSolutions/Neetcode150/pointer_problems/11_container-with-most-water.ts
function maxArea(height: number[]): number {
    let l = 0;
    let r = height.length - 1;
    let maxArea = 0;

    while (l < r) {
        const area = Math.min(height[l], height[r]) * (r - l);
        maxArea = Math.max(maxArea, area);

        if (height[l] < height[r]) {
            l++;
        }
        else {
            r--;
        }
    }
    
    return maxArea;
};